File: android/src/main/kotlin/com/snowplowanalytics/snowplow_tracker/SnowplowTrackerController.kt
// Copyright (c) 2022-present Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

package com.snowplowanalytics.snowplow_tracker

import android.content.Context

import com.snowplowanalytics.snowplow.Snowplow;
import com.snowplowanalytics.snowplow.configuration.Configuration;
import com.snowplowanalytics.snowplow.event.Event
import com.snowplowanalytics.snowplow_tracker.readers.configurations.DefaultTrackerConfiguration
import com.snowplowanalytics.snowplow_tracker.readers.messages.*

object SnowplowTrackerController {

    fun createTracker(messageReader: CreateTrackerMessageReader, context: Context) {
        val controllers: MutableList<Configuration> = mutableListOf()

        val networkConfigReader = messageReader.networkConfig;
        val networkConfiguration = networkConfigReader.toConfiguration()

        val trackerConfigReader = messageReader.trackerConfig
        if (trackerConfigReader == null) {
            controllers.add(DefaultTrackerConfiguration.toConfiguration(null, context))
        } else {
            controllers.add(trackerConfigReader.toConfiguration(context))
        }

        val subjectConfigReader = messageReader.subjectConfig
        subjectConfigReader?.let { controllers.add(it.toConfiguration()) }

        val gdprConfigReader = messageReader.gdprConfig
        gdprConfigReader?.let { controllers.add(it.toConfiguration()) }

        val emitterConfigReader = messageReader.emitterConfig
        emitterConfigReader?.let { controllers.add(it.toConfiguration()) }

        Snowplow.createTracker(
                context,
                messageReader.namespace,
                networkConfiguration,
                *controllers.toTypedArray()
        )
    }

    fun trackStructured(eventReader: EventMessageReader) {
        val structured = eventReader.toStructuredWithContexts()
        trackEvent(structured, eventReader)
    }

    fun trackSelfDescribing(eventReader: EventMessageReader) {
        val selfDescribing = eventReader.toSelfDescribingWithContexts()
        trackEvent(selfDescribing, eventReader)
    }

    fun trackScreenView(eventReader: EventMessageReader) {
        val screenView = eventReader.toScreenViewWithContexts()
        trackEvent(screenView, eventReader)
    }

    fun trackScrollChanged(eventReader: EventMessageReader) {
        val scroll = eventReader.toScrollChangedWithContexts()
        trackEvent(scroll, eventReader)
    }

    fun trackListItemView(eventReader: EventMessageReader) {
        val listItem = eventReader.toListItemViewWithContexts()
        trackEvent(listItem, eventReader)
    }

    fun trackTiming(eventReader: EventMessageReader) {
        val timing = eventReader.toTimingWithContexts()
        trackEvent(timing, eventReader)
    }

    fun trackConsentGranted(eventReader: EventMessageReader) {
        val consentGranted = eventReader.toConsentGrantedWithContexts()
        trackEvent(consentGranted, eventReader)
    }

    fun trackConsentWithdrawn(eventReader: EventMessageReader) {
        val consentWithdrawn = eventReader.toConsentWithdrawnWithContexts()
        trackEvent(consentWithdrawn, eventReader)
    }

    fun setUserId(messageReader: SetUserIdMessageReader) {
        val trackerController = Snowplow.getTracker(messageReader.tracker)

        trackerController?.subject?.userId = messageReader.userId
    }

    fun getSessionUserId(messageReader: GetParameterMessageReader): String? {
        val trackerController = Snowplow.getTracker(messageReader.tracker)

        return trackerController?.session?.userId
    }

    fun getSessionId(messageReader: GetParameterMessageReader): String? {
        val trackerController = Snowplow.getTracker(messageReader.tracker)

        return trackerController?.session?.sessionId
    }

    fun getSessionIndex(messageReader: GetParameterMessageReader): Int? {
        val trackerController = Snowplow.getTracker(messageReader.tracker)

        return trackerController?.session?.sessionIndex
    }

    fun startMediaTracking(messageReader: StartMediaTrackingMessageReader) {
        val trackerController = Snowplow.getTracker(messageReader.tracker)
        trackerController?.media?.startMediaTracking(
            messageReader.configuration.toConfiguration()
        )
    }

    fun endMediaTracking(messageReader: EndMediaTrackingMessageReader) {
        val trackerController = Snowplow.getTracker(messageReader.tracker)
        trackerController?.media?.endMediaTracking(messageReader.mediaTrackingId)
    }

    fun updateMediaTracking(messageReader: UpdateMediaTrackingMessageReader) {
        val trackerController = Snowplow.getTracker(messageReader.tracker)
        val mediaTracking = trackerController?.media?.getMediaTracking(messageReader.mediaTrackingId)
        mediaTracking?.update(
            player = messageReader.player?.toMediaPlayerEntity(),
            ad = messageReader.ad?.toMediaAdEntity(),
            adBreak = messageReader.adBreak?.toMediaAdBreakEntity()
        )
    }

    fun trackMediaAdBreakEndEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaAdBreakEndEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaAdBreakStartEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaAdBreakStartEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaAdClickEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaAdClickEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaAdCompleteEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaAdCompleteEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaAdFirstQuartileEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaAdFirstQuartileEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaAdMidpointEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaAdMidpointEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaAdPauseEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaAdPauseEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaAdResumeEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaAdResumeEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaAdSkipEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaAdSkipEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaAdStartEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaAdStartEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaAdThirdQuartileEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaAdThirdQuartileEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaBufferEndEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaBufferEndEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaBufferStartEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaBufferStartEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaEndEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaEndEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaErrorEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaErrorEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaFullscreenChangeEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaFullscreenChangeEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaPauseEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaPauseEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaPictureInPictureChangeEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaPictureInPictureChangeEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaPlayEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaPlayEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaPlaybackRateChangeEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaPlaybackRateChangeEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaQualityChangeEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaQualityChangeEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaReadyEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaReadyEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaSeekEndEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaSeekEndEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaSeekStartEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaSeekStartEvent()
        trackEvent(event, messageReader)
    }

    fun trackMediaVolumeChangeEvent(messageReader: EventMessageReader) {
        val event = messageReader.toMediaVolumeChangeEvent()
        trackEvent(event, messageReader)
    }

    private fun trackEvent(event: Event, messageReader: EventMessageReader) {
        val trackerController = Snowplow.getTracker(messageReader.tracker)
        val mediaTrackingId = messageReader.mediaTrackingId
        if (mediaTrackingId == null) {
            trackerController?.track(event)
        } else {
            val mediaTracking = trackerController?.media?.getMediaTracking(mediaTrackingId)
            mediaTracking?.track(
                event = event,
                player = messageReader.player?.toMediaPlayerEntity(),
                ad = messageReader.ad?.toMediaAdEntity(),
                adBreak = messageReader.adBreak?.toMediaAdBreakEntity()
            )
        }
    }

}

File: ios/Classes/SnowplowTrackerController.swift
// Copyright (c) 2022-present Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

import Foundation
import SnowplowTracker

class SnowplowTrackerController {
    // TODO: remove after 6.0.2 release of iOS tracker is published
    private static var mediaTrackings: [String: MediaTracking] = [:]

    static func createTracker(_ message: CreateTrackerMessageReader, arguments: [String: Any]) {
        var controllers: [ConfigurationProtocol] = []
        
        if let trackerConfig = message.trackerConfig {
            controllers.append(trackerConfig.toConfiguration())
        } else {
            controllers.append(TrackerConfigurationReader.defaultConfiguration())
        }
        if let subjectConfig = message.subjectConfig {
            controllers.append(subjectConfig.toConfiguration())
        }
        if let gdprConfig = message.gdprConfig {
            controllers.append(gdprConfig.toConfiguration())
        }
        if let emitterConfig = message.emitterConfig {
            controllers.append(emitterConfig.toConfiguration())
        }
        
        Snowplow.createTracker(
            namespace: message.namespace,
            network: message.networkConfig.toConfiguration(),
            configurations: controllers
        )
    }
    
    static func trackStructured(_ message: TrackStructuredMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = message.toStructured()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }
    
    static func trackSelfDescribing(_ message: TrackSelfDescribingMesssageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        if let event = message.toSelfDescribing(arguments: arguments) {
            trackEvent(event, eventMessage: eventMessage, arguments: arguments)
        }
    }
    
    static func trackScreenView(_ message: TrackScreenViewMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = message.toScreenView()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }
    
    static func trackScrollChanged(_ message: TrackScrollChangedMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = message.toScrollChanged()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }
    
    static func trackListItemView(_ message: TrackListItemViewMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = message.toListItemView()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }
    
    static func trackTiming(_ message: TrackTimingMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = message.toTiming()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }
    
    static func trackConsentGranted(_ message: TrackConsentGrantedMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = message.toConsentGranted()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }
    
    static func trackConsentWithdrawn(_ message: TrackConsentWithdrawnMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = message.toConsentWithdrawn()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }
    
    static func sessionUserId(_ message: GetParameterMessageReader) -> String? {
        return Snowplow.tracker(namespace: message.tracker)?.session?.userId
    }
    
    static func sessionId(_ message: GetParameterMessageReader) -> String? {
        return Snowplow.tracker(namespace: message.tracker)?.session?.sessionId
    }
    
    static func sessionIndex(_ message: GetParameterMessageReader) -> Int? {
        return Snowplow.tracker(namespace: message.tracker)?.session?.sessionIndex
    }
    
    static func setUserId(_ message: SetUserIdMessageReader) {
        let trackerController = Snowplow.tracker(namespace: message.tracker)
        trackerController?.subject?.userId = message.userId
    }

    static func startMediaTracking(_ message: StartMediaTrackingMessageReader, arguments: [String: Any]) {
        let configurationArguments = arguments["configuration"] as? [String: Any] ?? [:]
        let mediaTrackingConfiguration = message.configuration.toMediaTrackingConfiguration(arguments: configurationArguments)
        let trackerController = Snowplow.tracker(namespace: message.tracker)
        let mediaTracking = trackerController?.media.startMediaTracking(configuration: mediaTrackingConfiguration)
        self.mediaTrackings[mediaTrackingConfiguration.id] = mediaTracking
    }

    static func endMediaTracking(_ message: EndMediaTrackingMessageReader) {
        let trackerController = Snowplow.tracker(namespace: message.tracker)
        trackerController?.media.endMediaTracking(id: message.mediaTrackingId)
        self.mediaTrackings.removeValue(forKey: message.mediaTrackingId)
    }

    static func updateMediaTracking(_ message: UpdateMediaTrackingMessageReader) {
        let trackerController = Snowplow.tracker(namespace: message.tracker)
        if let media = self.mediaTrackings[message.mediaTrackingId] {
            media.update(
                player: message.player?.toMediaPlayerEntity(),
                ad: message.ad?.toMediaAdEntity(),
                adBreak: message.adBreak?.toMediaAdBreakEntity()
            )
        }
    }
    static func trackMediaAdBreakEndEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaAdBreakEndEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaAdBreakStartEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaAdBreakStartEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaAdClickEvent(eventMessage: TrackMediaAdEventMessageReader, arguments: [String: Any]) {
        let event = MediaAdClickEvent(percentProgress: message.eventData.percentProgress)
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaAdCompleteEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaAdCompleteEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaAdFirstQuartileEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaAdFirstQuartileEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaAdMidpointEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaAdMidpointEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaAdPauseEvent(_ message: TrackMediaAdEventMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaAdPauseEvent(percentProgress: message.eventData.percentProgress)
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaAdResumeEvent(_ message: TrackMediaAdEventMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaAdResumeEvent(percentProgress: message.eventData.percentProgress)
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaAdSkipEvent(_ message: TrackMediaAdEventMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaAdSkipEvent(percentProgress: message.eventData.percentProgress)
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaAdStartEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaAdStartEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaAdThirdQuartileEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaAdThirdQuartileEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaBufferEndEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaBufferEndEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaBufferStartEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaBufferStartEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaEndEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaEndEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaErrorEvent(_ message: TrackMediaErrorEventMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = message.toMediaErrorEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaFullscreenChangeEvent(_ message: TrackMediaFullscreenChangeEventMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = message.toMediaFullscreenChangeEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaPauseEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaPauseEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaPictureInPictureChangeEvent(_ message: TrackMediaPictureInPictureChangeEventMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = message.toMediaPictureInPictureChangeEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaPlayEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaPlayEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    // static func trackMediaPlaybackRateChangeEvent(_ message: TrackMediaPlaybackRateChangeEventMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
    //     let event = message.toMediaPlaybackRateChangeEvent()
    //     trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    // }

    static func trackMediaQualityChangeEvent(_ message: TrackMediaQualityChangeEventMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = message.toMediaQualityChangeEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaReadyEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaReadyEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaSeekEndEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaSeekEndEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaSeekStartEvent(eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = MediaSeekStartEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    static func trackMediaVolumeChangeEvent(_ message: TrackMediaVolumeChangeEventMessageReader, eventMessage: EventMessageReader, arguments: [String: Any]) {
        let event = message.toMediaVolumeChangeEvent()
        trackEvent(event, eventMessage: eventMessage, arguments: arguments)
    }

    private static func trackEvent(_ event: Event, eventMessage: EventMessageReader, arguments: [String: Any]) {
        eventMessage.addContextsToEvent(event, arguments: arguments)
        let trackerController = Snowplow.tracker(namespace: eventMessage.tracker)
        if let mediaTrackingId = eventMessage.mediaTrackingId {
            if let media = self.mediaTrackings[mediaTrackingId] {
                media.track(
                    event,
                    player: eventMessage.player?.toMediaPlayerEntity(),
                    ad: eventMessage.ad?.toMediaAdEntity(),
                    adBreak: eventMessage.adBreak?.toMediaAdBreakEntity()
                )
            }
        } else {
            trackerController?.track(event)
        }
    }
}
